Microsoft.Dynamics365.UIAutomation.Browser/Extensions/SeleniumExtensions.cs

```diff
@@ -555,107 +367,203 @@ public static object WaitForScript(this IWebDriver driver, string script, TimeSp
+            => WaitUntilVisible(driver, by, timeout, null, exceptionMessage);
+
+        public static IWebElement WaitUntilVisible(this ISearchContext driver, By by, 
+            TimeSpan? timeout, 
+            Action<IWebElement> successCallback, 
+            string exceptionMessage)
         {
-            return WaitUntilVisible(driver, by, timeout, null, null);
+            if (string.IsNullOrWhiteSpace(exceptionMessage)) 
+                exceptionMessage = $"Unable to find any visible element by: {by}";
+            return WaitUntilVisible(driver, by, timeout, successCallback, () => throw new InvalidOperationException(exceptionMessage));
         }
 
-        public static bool WaitUntilVisible(this IWebDriver driver, By by, TimeSpan timeout, Action<IWebDriver> successCallback)
+        public static IWebElement WaitUntilVisible(this ISearchContext driver, By by,
+            TimeSpan? timeout = null,
+            Action<IWebElement> successCallback = null,
+            Action failureCallback = null)
         {
-            return WaitUntilVisible(driver, by, timeout, successCallback, null);
+            return WaitUntil(driver, d => d.FindVisible(by), timeout, successCallback, failureCallback);
         }
 
-        public static bool WaitUntilVisible(this IWebDriver driver, By by, TimeSpan timeout, Action<IWebDriver> successCallback, Action<IWebDriver> failureCallback)
+
+        public static IWebElement WaitUntilClickable(this ISearchContext driver, By by, string exceptionMessage) 
+            => WaitUntilClickable(driver, by, null, null, exceptionMessage);
+        
+        public static IWebElement WaitUntilClickable(this ISearchContext driver, By by, Action<IWebElement> successCallback, string exceptionMessage) 
+            => WaitUntilClickable(driver, by, null, successCallback, exceptionMessage);
+
+        public static IWebElement WaitUntilClickable(this ISearchContext driver, By by, TimeSpan timeout, string exceptionMessage) 
+            => WaitUntilClickable(driver, by, timeout, null, exceptionMessage);
+
+        public static IWebElement WaitUntilClickable(this ISearchContext driver, By by, TimeSpan? timeout, Action<IWebElement> successCallback, string exceptionMessage)
         {
-            WebDriverWait wait = new WebDriverWait(driver, timeout);
-            bool? success;
+            if (string.IsNullOrWhiteSpace(exceptionMessage)) 
+                exceptionMessage = $"Unable to find any clickable element by: {by}";
+            return WaitUntilClickable(driver, by, timeout, successCallback, () => throw new InvalidOperationException(exceptionMessage));
+        }
+
+        public static IWebElement WaitUntilClickable(this ISearchContext driver, By by,
+            TimeSpan? timeout = null,
+            Action<IWebElement> successCallback = null,
+            Action failureCallback = null)
+        {
+            return WaitUntil(driver, d => d.FindClickable(by), timeout,
```

Comment:
@AngelRodriguez8008 this is throwing StaleElementReferenceException before timeout? any Clue?